### DISCUSSION TITLE
about how to destroy a QObject

### DISCUSSION THREAD

**Discussion by wangdaye078** [MeVisLab/pythonqt#110] [Q&A]
from PythonQt import QtCore
qObject = QtCore.QObject(parent)

i try to destroy qObject;
1.qObject = None
2.del qObject 
3.qObject.deleteLater()

only 3 i found ~QObject Executed

anyone know the other way to destroy qObject?

**Comment by wangdaye078** [comment]
i found
qObject = QtCore.QObject()
qObject = None

will destroy qObject

**Comment by mrbean-bremen** [comment]
The object will be destroyed after the Python wrapper is no longer referenced (which happens in your example where you assign `None` to the variable previously holding the object).
Generally, there should be no need to explicity destroy these objects - they (e.g. the wrapper objects) follow the same rules regarding object lifetime as any Python object.

**Comment by wangdaye078** [comment]
In fact, I am trying to dynamically load a script.

file.py like this

```py
from PythonQt import QtCore

class SomeClass():
　　def ▁▁init▁▁(self):
　　　　print("SomeClass __init__")
　　def ▁▁del▁▁(self):
　　　　print("SomeClass __del__")

someClass = SomeClass()

class ObjectManager():
　　def ▁▁init▁▁(self):
　　　　self.qObject = QtCore.QLabel(parent)
　　　　self.qObject.objectName = "someObj"
　　def ▁▁del▁▁(self):
　　　　self.qObject.deleteLater()

global objectManager

def doSome_slot():
　　pass

def main(init):
　　global objectManager
　　if (_connect):
　　　　objectManager = ObjectManager()
　　　　objectManager.qObject.connect(objectManager.qObject., "some_signal()", doSome_slot)
　　else:
　　　　objectManager.qObject.disconnect(objectManager.qObject., "some_signal()", doSome_slot)
　　　　objectManager = None
```
c++ like this
```cpp
PythonQtObjectPtr Init(QObject* parent)
{
　　PythonQt::init(PythonQt::IgnoreSiteModule | PythonQt::RedirectStdOut, QByteArray());
　　PythonQt_QtAll::init();
　　PythonQtObjectPtr module = PythonQt::self()->createModuleFromScript("ModuleName");
　　module.addObject("parent", parent);
　　module.evalFile("file.py");
　　module.call("main", QVariantList() << true);
　　return module;
}

void Finalize(PythonQtObjectPtr module)
{
　　module.call("main", QVariantList() << false);
　　//will print "SomeClass __del__"
　　_PyModule_Clear(module);
　　PythonQtObjectPtr dict = PyImport_GetModuleDict();
　　PyDict_DelItemString(dict, "ModuleName");
}
```
if i not write deleteLater, the qObject will always remain in memory;
if i run Init twice, the parent will have two children named "someObj"

**Comment by wangdaye078** [comment]
i getted
pythonqt add 4 methods to wrapper.
"className", "Return the classname of the object"
"inherits", "Returns if the class inherits or is of given type name"
"help", "Shows the help of available methods for this class"
"delete", "Deletes the C++ object (at your own risk, my friend!)"

so can write as "self.qObject.delete()"

**Comment by mrbean-bremen** [comment]
This behaves as expected with [Qt ownership rules](https://doc.qt.io/qt-6/objecttrees.html) - the parent object holds all if its children. To remove a child object you have to remove it from its parent instead of deleting it. If it is no longer held by another Qt object, or by a Python wrapper, it will be deleted. Deleting it yourself this way may even be dangerous.

**Comment by wangdaye078** [comment]
i tried

from PythonQt import QtCore, QtGui
qObj = QtGui.QLabel(parent)
....
qObj.setParent(None)
qObj = None

it will work fine

**Reply by mrbean-bremen** [reply]
Yes, `setParent(None)` should be correct. The second one (`qObj = None`) may not be needed, as `qObj` will get out of scope itself anyway, but it also doesn't do any harm.
